monkey/agent_plugins/exploiters/hadoop/src/plugin.py

```diff
@@ -0,0 +1,180 @@
+from infection_monkey.exploit.tools.helpers import get_agent_dst_path
+from infection_monkey.exploit.tools.http_tools import HTTPTools
+from infection_monkey.exploit.tools.web_tools import build_urls
+from infection_monkey.i_puppet import ExploiterResultData
+from infection_monkey.model import (
+    HADOOP_LINUX_COMMAND,
+    HADOOP_WINDOWS_COMMAND,
+    MONKEY_ARG,
+    TargetHost,
+)
+from infection_monkey.propagation_credentials_repository import IPropagationCredentialsRepository
+from infection_monkey.utils.commands import (
+    DROPPER_TARGET_PATH_LINUX,
+    DROPPER_TARGET_PATH_WIN64,
+    build_monkey_commandline,
+)
+
+from .hadoop import Hadoop
+from .hadoop_options import HadoopOptions
+
+logger = logging.getLogger(__name__)
+
+
+class Plugin:
+    _EXPLOITED_SERVICE = "Hadoop"
+
+    def __init__(
+        self,
+        plugin_name: str,
+        agent_id: UUID,
+        agent_event_publisher: IAgentEventPublisher,
+        agent_binary_repository: IAgentBinaryRepository,
+        propagation_credentials_repository: IPropagationCredentialsRepository,
+    ):
+        """
+        :param monkey_target_paths: Where to upload the monkey at the target host system.
+        Dict in format {'linux': '/tmp/monkey.sh', 'win64':... }
+        """
+        self._agent_id = agent_id
+        self._agent_event_publisher = agent_event_publisher
+        self._agent_binary_repository = agent_binary_repository
+        self.monkey_target_paths = {
+            "linux": DROPPER_TARGET_PATH_LINUX,
+            "windows": DROPPER_TARGET_PATH_WIN64,
+        }
+        self.vulnerable_urls = []
+        self.exploit_info = {
+            "display_name": self._EXPLOITED_SERVICE,
+            "started": "",
+            "finished": "",
+            "vulnerable_urls": [],
+            "vulnerable_ports": [],
+            "executed_cmds": [],
+        }
+        self.exploit_attempts = []
+        self.host = None
+        self.exploit_result = ExploiterResultData()
+        self.servers = []
+        self.hadoop_exploit = None
+
+    def run(
+        self,
+        host: TargetHost,
+        servers: Sequence[str],
+        current_depth: int,
+        options: Dict[str, Any],
+        interrupt: Event,
+    ) -> ExploiterResultData:
+        self.host = host
+        self.servers = servers
+        self.current_depth = current_depth
+
+        # HTTP ports options are hack because they are needed in fingerprinters
+        if "http_ports" in options:
+            del options["http_ports"]
+        self._options = HadoopOptions(**options)
+        self.hadoop_exploit = Hadoop(
+            host, self._options, interrupt, self._agent_id, self._agent_event_publisher
+        )
+
+        self._pre_exploit()
+        try:
+            return self._exploit_host()
+        except Exception as e:
+            logger.error("Exception in run", exc_info=True)
+            raise e
+        finally:
+            self._post_exploit()
```

Comment:
If you `raise e`, the host program has no way to retrieve `self.exploit_result` from this component. In other words, this finally is pointless.

Generally speaking, we can't raise exceptions from plugins to the host program. There's not really a mechanism to support that.